demo

```diff
@@ -62,16 +80,138 @@ if [[ "$?" -eq 1 ]]; then
+
+DOCKER_COMPOSE_PROJECT_NAME="${DOCKER_COMPOSE_PROJECT_NAME:-integration}"
+
+# Make sure that the demo environment is brought down on SIGINT
+exitfunc() {
+    retval=$(docker-compose \
+         -f docker-compose.yml \
+         -f docker-compose.storage.minio.yml \
+         -f docker-compose.demo.yml \
+         -p ${DOCKER_COMPOSE_PROJECT_NAME} \
+         $CLIENT \
+         down)
+    exit $retval
+}
+
+trap exitfunc SIGINT
+
+echo "Starting the Mender demo environment..."
+
+docker-compose \
+    -f docker-compose.yml \
+    -f docker-compose.storage.minio.yml \
+    -f docker-compose.demo.yml \
+    $CLIENT \
+    "$@" -d
+
+USER='mender-demo@example.com'
+PASSWORD=$(< /dev/urandom tr -dc A-Za-z0-9 | head -c12)
+
+MENDER_SERVER_USER="${USER}:${PASSWORD}"
+MENDER_SERVER_URI='https://localhost'
+
+# Block until the useradm service returns an HTTP response.
+# Which means that it is up.
+echo "Waiting for the useradm service to start up"
+curl --silent -k -X POST -u $MENDER_SERVER_USER\
+     --fail\
+     --retry-connrefused\
+     --retry 60\
+     --retry-delay 2\
+     --retry-max-time 120\
+     $MENDER_SERVER_URI/api/management/v1/useradm/auth/login > /dev/null
+
+retval=$?
+case $retval in
+    0)  ;; # User exists - continue.
+    22) ;; # Server 400 error, ie, server is up - continue.
+    *) echo "Failed to access the mender-useradm service"
+       echo "Please check the docker-compose setup."
+       echo "curl exit code" $retval
+       exit $retval
+       ;;
+esac
+
+echo "Creating a new user..."
+
+# Create a new user.
+docker exec ${DOCKER_COMPOSE_PROJECT_NAME}_mender-useradm_1 /usr/bin/useradm create-user --username=${USER} --password=${PASSWORD}
+
+retval=$?
+case $retval in
+     0) echo "New user created." ;;
+     5) echo "The user already exists"; exit 1;; # No need to continue if a demo user exists
+     *) exit $retval ;;
+esac
+
+echo "Getting JWT Token..."
+
+JWT=$(curl --silent -k -X POST -u $MENDER_SERVER_USER\
+           --fail\
+           --retry-connrefused\
+           --retry 60\
+           --retry-delay 2\
+           --retry-max-time 120\
+           $MENDER_SERVER_URI/api/management/v1/useradm/auth/login)
+
+retval=$?
+if [ $retval -ne 0 ]; then
+    echo "Failed to get the 'JWT' token from the useradm service."
+    echo "This is needed in order to use the demo Artifact."
+    exit $retval
+fi
+
+echo "Uploading the artifact to the server..."
+errout=$(curl -k --request POST\
+              --silent\
+              --show-error\
+              --retry-connrefused\
+              --retry-delay 2\
+              --retry 60 \
+              --retry-max-time 120\
+              --header "Authorization: Bearer ${JWT}"\
+              --form "size=${ARTIFACT_SIZE_BYTES}"\
+              --form "artifact=@${DEMO_ARTIFACT_NAME}"\
+              "${MENDER_SERVER_URI}/api/management/v1/deployments/artifacts" | jq .error)
+
+retval=$?
```

Comment:
The retval here is for the `jq .error` command, I don't think that is your intention. You coud set this `pipefail` flag, but I don't know if that is POSIX standard or not. @kacf ?

Reply:
I will split this up into two parts. And handle the errors independently. `pipefail` is a bashism btw ;)

Reply:
Indeed, good catch. The `curl` error will go unnoticed here. Feel free to use `pipefail` and other bashisms, the script is explicitly called with `bash`, not `sh`.